<img height="180px" src="https://drive.google.com/uc?export=view&id=141XOz6N4nk8Ru1sAl7vOsAToCLrSFCAX" alt="SDA logo" align="left" hspace="30px" vspace="50px"/>

# Welcome to your next notebook with SDA!

During the classes we will mostly use [Google Colaboratory](https://colab.research.google.com/?hl=en) which is a free Jupyter notebook environment that requires no setup and runs entirely in the cloud.

However, for bigger projects, especially involving Deep Learning and/or big data reading, it might be a better choice to setup Jupyter Notebook or Jupyter Lab on your computer. Also, it is worth noticing that there is a great number of useful extensions (see [nbextensions](https://jupyter-contrib-nbextensions.readthedocs.io/en/latest/index.html) and [jupyter-labextension](https://jupyterlab.readthedocs.io/en/stable/user/extensions.html)) not available for Colab users.

<img src="https://drive.google.com/uc?export=view&id=1UO2urRciECzoKE_vHy4RMGfFbkOWOGlW" alt="SDA logo" align="left" width="100px" hspace="10px" vspace="10px"/>

# AI Engineer
## Module 2: Data Processing and AI/ML Models
### Data Processing with Pandas

After the **<font color='#ed7d31'>Data Processing with Pandas</font>** course you will dive deeper into Python and familiarize yourself with concepts especially useful for future Data Analysts, Data Scientists and AI Engineers:
* data cleaning,
* data manipulation
* basic data analysis,
* Pandas data structures,
  * `pd.Series()`
  * `pd.DataFrame()`
* Pandas functions and methods.

<img src="https://drive.google.com/uc?export=view&id=1UO2urRciECzoKE_vHy4RMGfFbkOWOGlW" alt="SDA logo" align="left" width="100px" hspace="10px" vspace="10px"/>
<br>

# Operations on files

<br><br>


#### **<font color='#306998'>MOUNT </font><font color='#ffd33b'>GOOGLE DRIVE</font>**

Download `sentences.txt` file from the following kaggle site:<br>
https://www.kaggle.com/datasets/olgabelitskaya/toy-data-for-text-processing

Upload it to your Google Drive, mount it and load the data.

*Hint: you can copy the path to the file or change the working directory and then use just the file name with .txt extenstion.*

In [1]:
# from google.colab import drive
# drive.mount('/content/drive')

#### **<font color='#306998'>TASK </font><font color='#ffd33b'>FOR YOU</font>**

Download `chipotle.tsv` file from the following kaggle site:
https://www.kaggle.com/datasets/navneethc/chipotle

Upload it to your Google Drive, mount it and load the data.

*Hint: you can copy the path to the file or change the working directory and then use just the file name with .tsv extenstion.*

In [2]:
import pandas as pd

In [3]:
chipotle = pd.read_csv("../dane/chipotle.tsv", sep="\t")

### Overwriting data

You can also overwrite or append data with the assignment command `=`:
```python
df[“column”] = new_data_series
df[[“column”]] = new_dataframe
df[[“column1”, “column2”]] = new_dataframe
df.loc[column:row] = new_dataframe
df.iloc[column_number:row_number] = new_dataframe
```

In [4]:
chipotle['new_col'] = 'oh'

In [5]:
chipotle.head()

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description item_price new_col  
0                                                NaN     $2.39       oh  
1                                       [Clementine]     $3.39       oh  
2                                            [Apple]     $3.39       oh  
3                                                NaN     $2.39       oh  
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98       oh

In [6]:
chipotle = chipotle.drop(columns=['new_col'])

<img src="https://drive.google.com/uc?export=view&id=141XOz6N4nk8Ru1sAl7vOsAToCLrSFCAX" alt="SDA logo" width="150" align='right'/>

## Operating on data

### .apply()

All the following operations for the dataframe will return values for each of the columns, we can select the column for which we want to perform these actions:
> `df.apply(function)` - applying a function to each column or cell in the set

In [7]:
chipotle['item_price [$]'] = chipotle['item_price'].apply(lambda x: float(x[1:]))

In [8]:
chipotle['item_price [$]'].apply(round)
# chipotle['item_price [$]'].apply(lambda x: round(x))

0        2
1        3
2        3
3        2
4       17
        ..
4617    12
4618    12
4619    11
4620     9
4621     9
Name: item_price [$], Length: 4622, dtype: int64

**<font color='#5b9bd5'>Please see [TASK 3](#scrollTo=WKJVnZD9iAyI).</font>**

### .groupby()

You need to segment your database from time to time. In addition to computing statistics for all values, sometimes these values can be grouped. In pandas, the `.groupby()` method is used for this purpose.
```python
df.groupby(“column”).operation().column
```

In [9]:
chipotle.groupby('item_name')

In [10]:
chipotle[['item_name', 'item_price [$]', 'quantity']].groupby('item_name').max().head()

item_price [$]  quantity
item_name                                      
6 Pack Soft Drink               12.98         2
Barbacoa Bowl                   11.75         1
Barbacoa Burrito                11.75         1
Barbacoa Crispy Tacos           18.50         2
Barbacoa Salad Bowl             11.89         1

In [11]:
chipotle[chipotle['item_name'] == "6 Pack Soft Drink"].head()

order_id  quantity          item_name choice_description item_price  \
298       129         1  6 Pack Soft Drink           [Sprite]     $6.49    
341       148         1  6 Pack Soft Drink        [Diet Coke]     $6.49    
357       154         1  6 Pack Soft Drink             [Coke]     $6.49    
388       168         1  6 Pack Soft Drink        [Diet Coke]     $6.49    
417       182         1  6 Pack Soft Drink        [Diet Coke]     $6.49    

     item_price [$]  
298            6.49  
341            6.49  
357            6.49  
388            6.49  
417            6.49

### Removing data

The df.drop() command is used to delete data - as a parameter we give a list of indexes or columns to be deleted and axis - whether an index or a column should be deleted.



In [12]:
chipotle.drop(columns='item_price')
# chipotle.drop('item_price', axis=1)

order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
1            1         1                                   Izze   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
4            2         2                           Chicken Bowl   
...        ...       ...                                    ...   
4617      1833         1                          Steak Burrito   
4618      1833         1                          Steak Burrito   
4619      1834         1                     Chicken Salad Bowl   
4620      1834         1                     Chicken Salad Bowl   
4621      1834         1                     Chicken Salad Bowl   

                                     choice_description  item_price [$]  
0                                                   NaN            2.39  
1                                          [Clementine]            3.39  
2                                               [Apple]            3.39  
3                                                   NaN            2.39  
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...           16.98  
...                                                 ...             ...  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...           11.75  
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...           11.75  
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...           11.25  
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...            8.75  
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...            8.75  

[4622 rows x 5 columns]

In [13]:
# chipotle = chipotle.drop(columns='item_price')
# or
chipotle.drop(columns='item_price', inplace=True)

We can also remove incomplete data (NaNy) with the df.dropna() command - the axis parameter removes columns or rows.



In [14]:
chipotle.dropna()

order_id  quantity           item_name  \
1            1         1                Izze   
2            1         1    Nantucket Nectar   
4            2         2        Chicken Bowl   
5            3         1        Chicken Bowl   
7            4         1       Steak Burrito   
...        ...       ...                 ...   
4617      1833         1       Steak Burrito   
4618      1833         1       Steak Burrito   
4619      1834         1  Chicken Salad Bowl   
4620      1834         1  Chicken Salad Bowl   
4621      1834         1  Chicken Salad Bowl   

                                     choice_description  item_price [$]  
1                                          [Clementine]            3.39  
2                                               [Apple]            3.39  
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...           16.98  
5     [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...           10.98  
7     [Tomatillo Red Chili Salsa, [Fajita Vegetables...           11.75  
...                                                 ...             ...  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...           11.75  
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...           11.75  
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...           11.25  
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...            8.75  
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...            8.75  

[3376 rows x 5 columns]

In [15]:
len(chipotle), len(chipotle.dropna())

(4622, 3376)

### Merging data

In fact, often instead of using one large database, we connect many smaller ones (it is easier to manage them, avoid redundancy, additionally we save disk space and achieve higher speed). Moreover, often the data comes from various sources, hence is splitted naturally betwenn files.

We can combine the data in the library in a few ways:
* `df.append(object)` method adds new rows to the end of the existing dataframe **DEPRECATED Use `pd.concat()` instead**
* `df.merge()` method, which in its assumptions is very similar to `SQL JOIN` - you can choose the joining method - inner (common), outer (sum), left, right and specify the `on` parameter - the name of the column to be a hyphen.

Additionally, we can give the new column name ourselves and thus add the data to the existing data: df ["new column"] = data

There are also other methods, such as, `pd.concat()` and `pd.join()`, see more https://pandas.pydata.org/docs/user_guide/merging.html.

In [16]:
import pandas as pd

Inspired by https://www.w3resource.com/python-exercises/pandas/joining-and-merging/index.php.



In [17]:
test_0_group_0 = pd.DataFrame([
    [0, "S1", "Danniella Fenton", 200],
    [1, "S2", "Ryder Storey", 210],
    [2, "S3", "Bryce Jensen", 190],
    [3, "S4", "Ed Bernal", 222],
    [4, "S5", "Kwame Morin", 199]
], columns=["id", "student_id", "name", "marks"]).set_index("id")

test_0_group_0

student_id              name  marks
id                                    
0          S1  Danniella Fenton    200
1          S2      Ryder Storey    210
2          S3      Bryce Jensen    190
3          S4         Ed Bernal    222
4          S5       Kwame Morin    199

In [18]:
test_0_group_1 = pd.DataFrame([
    [0, "S6", "Scarlette Fisher", 201],
    [1, "S7", "Carla Williamson", 200],
    [2, "S8", "Dante Morse", 198],
    [3, "S9", "Kaiser William", 219],
    [4, "S10", "Madeeha Preston", 201]
], columns=["id", "student_id", "name", "marks"]).set_index("id")

test_0_group_1

student_id              name  marks
id                                    
0          S6  Scarlette Fisher    201
1          S7  Carla Williamson    200
2          S8       Dante Morse    198
3          S9    Kaiser William    219
4         S10   Madeeha Preston    201

In [19]:
test_0 = pd.concat([test_0_group_0, test_0_group_1], ignore_index=True)
test_0

student_id              name  marks
0         S1  Danniella Fenton    200
1         S2      Ryder Storey    210
2         S3      Bryce Jensen    190
3         S4         Ed Bernal    222
4         S5       Kwame Morin    199
5         S6  Scarlette Fisher    201
6         S7  Carla Williamson    200
7         S8       Dante Morse    198
8         S9    Kaiser William    219
9        S10   Madeeha Preston    201

Use the following data to create additional `test_0_group_2` with results from group 2:

```
  student_id          name  marks
5         S9    Liam Smith    205
6        S10  Olivia Jones    207
7        S11   Noah Taylor    195
8        S12    Emma Brown    210
9        S13     Ava Davis    202
```

Expand `test_0` to include those results.

In [20]:
test_1_group_0 = pd.DataFrame([
    [0, 215],
    [1, 185],
    [2, 160],
    [3, 230],
    [4, 198]
], columns=["id", "marks_1"]).set_index("id")

test_1_group_0

marks_1
id         
0       215
1       185
2       160
3       230
4       198

In [21]:
pd.concat([test_0_group_0, test_1_group_0], axis=1)

student_id              name  marks  marks_1
id                                             
0          S1  Danniella Fenton    200      215
1          S2      Ryder Storey    210      185
2          S3      Bryce Jensen    190      160
3          S4         Ed Bernal    222      230
4          S5       Kwame Morin    199      198

In [22]:
test_1_group_0 = pd.DataFrame([
    [0, "S4", "Ed Bernal", 230],
    [1, "S1", "Danniella Fenton", 215],
    [2, "S5", "Kwame Morin", 198],
    [3, "S2", "Ryder Storey", 185],
    [4, "S3", "Bryce Jensen", 160]
], columns=["id", "student_id", "name", "marks"]).set_index("id")

test_1_group_0

student_id              name  marks
id                                    
0          S4         Ed Bernal    230
1          S1  Danniella Fenton    215
2          S5       Kwame Morin    198
3          S2      Ryder Storey    185
4          S3      Bryce Jensen    160

In [23]:
pd.concat([test_0_group_0, test_1_group_0], axis=1)  # incorrect concatenation

student_id              name  marks student_id              name  marks
id                                                                        
0          S1  Danniella Fenton    200         S4         Ed Bernal    230
1          S2      Ryder Storey    210         S1  Danniella Fenton    215
2          S3      Bryce Jensen    190         S5       Kwame Morin    198
3          S4         Ed Bernal    222         S2      Ryder Storey    185
4          S5       Kwame Morin    199         S3      Bryce Jensen    160

In [24]:
tests_group_0 = pd.merge(
    left=test_0_group_0,
    right=test_1_group_0,
    on=["student_id", "name"],
    suffixes=["_test_0", "_test_1"]
)

tests_group_0

student_id              name  marks_test_0  marks_test_1
0         S1  Danniella Fenton           200           215
1         S2      Ryder Storey           210           185
2         S3      Bryce Jensen           190           160
3         S4         Ed Bernal           222           230
4         S5       Kwame Morin           199           198

### Sorting data

You can also use a method `df.sort_values(by="column name")`
with `ascending` parameter set to `True` for ascending order and `False` for descending one.

Moreover, often, after sorting the dataset, we want to drop the old indexes - the `df.reset_index(drop=True)` is used for this.

In [25]:
chipotle.sort_values('quantity')

order_id  quantity                     item_name  \
0            1         1  Chips and Fresh Tomato Salsa   
3020      1201         1                   Canned Soda   
3021      1201         1                 Bottled Water   
3022      1201         1                 Side of Chips   
3023      1202         1                  Chicken Bowl   
...        ...       ...                           ...   
2441       970         5                 Bottled Water   
3599      1443         7                 Bottled Water   
3887      1559         8                 Side of Chips   
4152      1660        10                 Bottled Water   
3598      1443        15  Chips and Fresh Tomato Salsa   

                                     choice_description  item_price [$]  
0                                                   NaN            2.39  
3020                                  [Diet Dr. Pepper]            1.09  
3021                                                NaN            1.09  
3022                                                NaN            1.69  
3023  [[Fresh Tomato Salsa (Mild), Roasted Chili Cor...            8.49  
...                                                 ...             ...  
2441                                                NaN            7.50  
3599                                                NaN           10.50  
3887                                                NaN           13.52  
4152                                                NaN           15.00  
3598                                                NaN           44.25  

[4622 rows x 5 columns]

In [26]:
chipotle.sort_values('item_price [$]', ascending=False)

order_id  quantity                     item_name  \
3598      1443        15  Chips and Fresh Tomato Salsa   
3480      1398         3                 Carnitas Bowl   
1254       511         4               Chicken Burrito   
3602      1443         4               Chicken Burrito   
3601      1443         3                Veggie Burrito   
...        ...       ...                           ...   
3936      1578         1                   Canned Soda   
2922      1162         1                 Bottled Water   
1396       567         1                   Canned Soda   
2562      1014         1                   Canned Soda   
1457       591         1                   Canned Soda   

                                     choice_description  item_price [$]  
3598                                                NaN           44.25  
3480  [Roasted Chili Corn Salsa, [Fajita Vegetables,...           35.25  
1254  [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...           35.00  
3602  [Fresh Tomato Salsa, [Rice, Black Beans, Chees...           35.00  
3601  [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...           33.75  
...                                                 ...             ...  
3936                                  [Diet Dr. Pepper]            1.09  
2922                                                NaN            1.09  
1396                                        [Coca Cola]            1.09  
2562                                        [Coca Cola]            1.09  
1457                                           [Sprite]            1.09  

[4622 rows x 5 columns]

In [27]:
chipotle.sort_values('item_price [$]', ascending=False).reset_index()  # inplace & drop

index  order_id  quantity                     item_name  \
0      3598      1443        15  Chips and Fresh Tomato Salsa   
1      3480      1398         3                 Carnitas Bowl   
2      1254       511         4               Chicken Burrito   
3      3602      1443         4               Chicken Burrito   
4      3601      1443         3                Veggie Burrito   
...     ...       ...       ...                           ...   
4617   3936      1578         1                   Canned Soda   
4618   2922      1162         1                 Bottled Water   
4619   1396       567         1                   Canned Soda   
4620   2562      1014         1                   Canned Soda   
4621   1457       591         1                   Canned Soda   

                                     choice_description  item_price [$]  
0                                                   NaN           44.25  
1     [Roasted Chili Corn Salsa, [Fajita Vegetables,...           35.25  
2     [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...           35.00  
3     [Fresh Tomato Salsa, [Rice, Black Beans, Chees...           35.00  
4     [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...           33.75  
...                                                 ...             ...  
4617                                  [Diet Dr. Pepper]            1.09  
4618                                                NaN            1.09  
4619                                        [Coca Cola]            1.09  
4620                                        [Coca Cola]            1.09  
4621                                           [Sprite]            1.09  

[4622 rows x 6 columns]

### GroupBy & apply


In [28]:
chipotle

order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
1            1         1                                   Izze   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
4            2         2                           Chicken Bowl   
...        ...       ...                                    ...   
4617      1833         1                          Steak Burrito   
4618      1833         1                          Steak Burrito   
4619      1834         1                     Chicken Salad Bowl   
4620      1834         1                     Chicken Salad Bowl   
4621      1834         1                     Chicken Salad Bowl   

                                     choice_description  item_price [$]  
0                                                   NaN            2.39  
1                                          [Clementine]            3.39  
2                                               [Apple]            3.39  
3                                                   NaN            2.39  
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...           16.98  
...                                                 ...             ...  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...           11.75  
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...           11.75  
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...           11.25  
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...            8.75  
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...            8.75  

[4622 rows x 5 columns]

In [29]:
chipotle.groupby('item_name').sum()[['quantity', 'item_price [$]']]

quantity  item_price [$]
item_name                                                      
6 Pack Soft Drink                            55          356.95
Barbacoa Bowl                                66          672.36
Barbacoa Burrito                             91          894.75
Barbacoa Crispy Tacos                        12          120.21
Barbacoa Salad Bowl                          10          106.40
Barbacoa Soft Tacos                          25          250.46
Bottled Water                               211          302.56
Bowl                                          4           29.60
Burrito                                       6           44.40
Canned Soda                                 126          137.34
Canned Soft Drink                           351          438.75
Carnitas Bowl                                71          736.71
Carnitas Burrito                             60          597.83
Carnitas Crispy Tacos                         8           77.96
Carnitas Salad                                1            8.99
Carnitas Salad Bowl                           6           66.34
Carnitas Soft Tacos                          40          375.94
Chicken Bowl                                761         7342.73
Chicken Burrito                             591         5575.82
Chicken Crispy Tacos                         50          472.13
Chicken Salad                                 9           81.09
Chicken Salad Bowl                          123         1228.75
Chicken Soft Tacos                          120         1108.09
Chips                                       230          494.34
Chips and Fresh Tomato Salsa                130          361.36
Chips and Guacamole                         506         2201.04
Chips and Mild Fresh Tomato Salsa             1            3.00
Chips and Roasted Chili Corn Salsa           23           67.85
Chips and Roasted Chili-Corn Salsa           18           43.02
Chips and Tomatillo Green Chili Salsa        45          132.75
Chips and Tomatillo Red Chili Salsa          50          147.50
Chips and Tomatillo-Green Chili Salsa        33           78.87
Chips and Tomatillo-Red Chili Salsa          25           59.75
Crispy Tacos                                  2           14.80
Izze                                         20           67.80
Nantucket Nectar                             29           98.31
Salad                                         2           14.80
Side of Chips                               110          185.90
Steak Bowl                                  221         2260.19
Steak Burrito                               386         3851.43
Steak Crispy Tacos                           36          357.34
Steak Salad                                   4           35.66
Steak Salad Bowl                             31          343.59
Steak Soft Tacos                             56          536.05
Veggie Bowl                                  87          867.99
Veggie Burrito                               97          934.77
Veggie Crispy Tacos                           1            8.49
Veggie Salad                                  6           50.94
Veggie Salad Bowl                            18          182.50
Veggie Soft Tacos                             8           73.96

### Split text

Useful when you have multiple traits in one column.

In [30]:
chipotle['item_name'].str.split(' ').map(lambda x: x[0])

0           Chips
1            Izze
2       Nantucket
3           Chips
4         Chicken
          ...    
4617        Steak
4618        Steak
4619      Chicken
4620      Chicken
4621      Chicken
Name: item_name, Length: 4622, dtype: object

In [31]:
chipotle[chipotle['item_name'].str.startswith('Chicken')]['item_name'].value_counts()

item_name
Chicken Bowl            726
Chicken Burrito         553
Chicken Soft Tacos      115
Chicken Salad Bowl      110
Chicken Crispy Tacos     47
Chicken Salad             9
Name: count, dtype: int64

> Fun fact<br>
The process of obtaining, preprocessing data and extracting features from them takes up to 80% of a data engineer's work time.

<img src="https://drive.google.com/uc?export=view&id=1UO2urRciECzoKE_vHy4RMGfFbkOWOGlW" alt="SDA logo" align="left" width="100px" hspace="10px" vspace="10px"/>
<br>

# TASKS

## **<font color='#306998'>TASK </font><font color='#ffd33b'>11</font>**

Re-create a dataframe from [TASK 1](https://colab.research.google.com/drive/1uP23sSdiDuRcnquO7zxxM2reAupO_N1H#scrollTo=z5eXe9__yhZ6&line=9&uniqifier=1)  (with the 10 names of pupils and the exam points).

1. Add a third column to the frame which will contain the percentage of student scores on this exam (that is, each numerical value must be divided by the maximum score that could be obtained on this exam).

1. Next, we want to anonymize the first name column: we want only the first and last letters of the first name to be left. For example, for the name "John" we want to keep "J ... n".


In [32]:
data = {
    "Pupil": ["Sebastian", "Ela", "Celina", "Jolanta", "Izabela", "Dariusz", "Szymon", "Dawid", "Wiktoria", "Weronika"],
    "Points": [10, 12, 22, 20, 20, 9, 8, 21, 11, 14]
}
df = pd.DataFrame(data)

In [33]:
df

Pupil  Points
0  Sebastian      10
1        Ela      12
2     Celina      22
3    Jolanta      20
4    Izabela      20
5    Dariusz       9
6     Szymon       8
7      Dawid      21
8   Wiktoria      11
9   Weronika      14

In [34]:
max_points = df["Points"].max()
max_points

22

In [35]:
df["score%"] = round(df["Points"] / max_points * 100, 2)

In [36]:
df

Pupil  Points  score%
0  Sebastian      10   45.45
1        Ela      12   54.55
2     Celina      22  100.00
3    Jolanta      20   90.91
4    Izabela      20   90.91
5    Dariusz       9   40.91
6     Szymon       8   36.36
7      Dawid      21   95.45
8   Wiktoria      11   50.00
9   Weronika      14   63.64

In [37]:
def coding(pupil):
  anonymized = pupil[0] + (len(pupil)-2) * "." + pupil[-1]
  return anonymized


In [38]:
df["Pupil"] = df["Pupil"].apply(coding)

In [39]:
df

Pupil  Points  score%
0  S.......n      10   45.45
1        E.a      12   54.55
2     C....a      22  100.00
3    J.....a      20   90.91
4    I.....a      20   90.91
5    D.....z       9   40.91
6     S....n       8   36.36
7      D...d      21   95.45
8   W......a      11   50.00
9   W......a      14   63.64

## **<font color='#306998'>TASK </font><font color='#ffd33b'>12</font>**

Sort the data frame by the number of points obtained for the exam, from highest to lowest. In case of the same number of points, the persons should be displayed in alphabetical order.

In [40]:
df.sort_values(by=["Points", "Pupil"], ascending=[False, True])

Pupil  Points  score%
2     C....a      22  100.00
7      D...d      21   95.45
4    I.....a      20   90.91
3    J.....a      20   90.91
9   W......a      14   63.64
1        E.a      12   54.55
8   W......a      11   50.00
0  S.......n      10   45.45
5    D.....z       9   40.91
6     S....n       8   36.36

## **<font color='#306998'>TASK </font><font color='#ffd33b'>13</font>**

Join the two dataframes defined below:
```python
student_data1 = pd.DataFrame({
    'student_id': ['S1', 'S2', 'S3', 'S4', 'S5'],
    'name': ['Danniella Fenton', 'Ryder Storey', 'Bryce Jensen', 'Ed Bernal', 'Kwame Morin']
})
```
together with:
```python
student_data2 = pd.DataFrame({
    'student_id': ['S6', 'S7', 'S8', 'S9', 'S10'],
    'name': ['Scarlette Fisher', 'Carla Williamson', 'Dante Morse', 'Kaiser William', 'Madeeha Preston']
})
```

In [41]:
student_data1 = pd.DataFrame({
    'student_id': ['S1', 'S2', 'S3', 'S4', 'S5'],
    'name': ['Danniella Fenton', 'Ryder Storey', 'Bryce Jensen', 'Ed Bernal', 'Kwame Morin']
})

In [42]:
student_data2 = pd.DataFrame({
    'student_id': ['S6', 'S7', 'S8', 'S9', 'S10'],
    'name': ['Scarlette Fisher', 'Carla Williamson', 'Dante Morse', 'Kaiser William', 'Madeeha Preston']
})

In [43]:
student_data = pd.concat([student_data1, student_data2], ignore_index=True)

In [44]:
student_data

student_id              name
0         S1  Danniella Fenton
1         S2      Ryder Storey
2         S3      Bryce Jensen
3         S4         Ed Bernal
4         S5       Kwame Morin
5         S6  Scarlette Fisher
6         S7  Carla Williamson
7         S8       Dante Morse
8         S9    Kaiser William
9        S10   Madeeha Preston

## **<font color='#306998'>TASK </font><font color='#ffd33b'>14</font>**

To the box created in the previous task, add the information about the students' results from the box below.
```python
exam_results = pd.DataFrame({
    'student_id': ['S2', 'S10', 'S3', 'S1', 'S7', 'S9', 'S5', 'S4', 'S8', 'S6'],
    'marks': [200, 210, 190, 222, 199, 201, 200, 198, 219, 201]
})
```

In [45]:
exam_results = pd.DataFrame({
    'student_id': ['S2', 'S10', 'S3', 'S1', 'S7', 'S9', 'S5', 'S4', 'S8', 'S6'],
    'marks': [200, 210, 190, 222, 199, 201, 200, 198, 219, 201]
})

In [46]:
student_data = pd.merge(left=student_data, right=exam_results, on=["student_id"])

In [47]:
student_data

student_id              name  marks
0         S1  Danniella Fenton    222
1         S2      Ryder Storey    200
2         S3      Bryce Jensen    190
3         S4         Ed Bernal    198
4         S5       Kwame Morin    200
5         S6  Scarlette Fisher    201
6         S7  Carla Williamson    199
7         S8       Dante Morse    219
8         S9    Kaiser William    201
9        S10   Madeeha Preston    210

## **<font color='#306998'>Information for TASKs </font><font color='#ffd33b'>15+</font>**

The following data was made available on:</br> https://www.kaggle.com/datasets/davidbnn92/weather-data-for-covid19-data-analysis/.

Try downloading it from there using the `Download` button or with the API command
```python
!kaggle datasets download -d davidbnn92/weather-data-for-covid19-data-analysis
```
(see https://www.kaggle.com/discussions/general/74235 for more details).

If the data is no longer available you can always download it from our Google Drive https://drive.google.com/drive/folders/1KXr6yUW7rE0LzUzuuMehEzam8D9zxGpv?usp=sharing.

### About Dataset

The dataset contains selected metereological features, such as temperature or wind speed, and was imported from the `NOAA GSOD dataset`, continuously updated to include recent measurments.

> Among others, you can find the following columns here:
* `Id`
* `Country/Region`
* `Date`
* `temp`: Mean temperature for the day in degrees Fahrenheit to tenths.
* `max`: Maximum temperature reported during the day.
* `min`: Minimum temperature reported during the day.
* `stp`: Mean station pressure for the day in millibars to tenths.
* `slp`: Mean sea level pressure for the day.
* `dewp`: Mean dew point for the day in [Fahrenheit to tenths].
* `wdsp`: Mean wind speed for the day in [knots to tenths].
* `prcp`: Total precipitation (rain and/or melted snow) reported during the day in [inches and hundredths]; `.00` indicates no measurable precipitation (includes a trace).
* `fog`: Indicators (1 = yes, 0 = no/not reported) for the occurrence during the day.
>
> Note that time of max/min temperatures varies by country and region, so this will sometimes not be the max for the calendar day.

## **<font color='#306998'>TASK </font><font color='#ffd33b'>15</font>**

Search through `training_data_with_weather_info_week_4.csv` file for weather data. Save the data frame as the variable `df_weather`.

In [48]:
df_weather = pd.read_csv("../dane/training_data_with_weather_info_week_4.csv", usecols=["Id", "Country_Region", "Date", "temp", "max", "min", "stp", "slp", "dewp", "wdsp", "prcp", "fog"])

In [49]:
df_weather.head()

Id Country_Region        Date  temp   min   max    stp     slp  dewp  wdsp  \
0   1    Afghanistan  2020-01-22  42.6  33.6  54.9  999.9  1024.3  27.4   9.4   
1   2    Afghanistan  2020-01-23  42.0  32.7  55.9  999.9  1020.8  22.8  14.9   
2   3    Afghanistan  2020-01-24  40.1  36.9  43.2  999.9  1018.6  34.5  10.4   
3   4    Afghanistan  2020-01-25  46.0  37.9  56.3  999.9  1018.0  37.8   6.1   
4   5    Afghanistan  2020-01-26  42.8  36.1  53.1  999.9  1014.8  33.2  10.8   

    prcp  fog  
0   0.00    0  
1  99.99    1  
2   0.17    1  
3   0.57    1  
4   0.00    1

## **<font color='#306998'>TASK </font><font color='#ffd33b'>16</font>**

Enter a new column `average monthly temperature (C)` containing the average temperature in degrees Celsius for each Country/Region.
* Round the result to one decimal place.
* Determine the average temperatures recorded in Togo over the years, and determine the lowest and highest value.
* Find 5 countries with the lowest average temps in March.


In [50]:
average_temp = df_weather[["Country_Region", "temp"]].groupby("Country_Region").mean().round(1)

In [51]:
average_temp

temp
Country_Region          
Afghanistan         35.6
Albania             52.6
Algeria             70.3
Andorra             34.7
Angola              82.6
...                  ...
Vietnam             77.7
West Bank and Gaza  52.5
Western Sahara      66.5
Zambia              72.3
Zimbabwe            70.0

[184 rows x 1 columns]

In [52]:
df_weather = pd.merge(df_weather, average_temp, on="Country_Region", suffixes=["", "average monthly temperature (C)"])

In [53]:
df_weather.head()

Id Country_Region        Date  temp   min   max    stp     slp  dewp  wdsp  \
0   1    Afghanistan  2020-01-22  42.6  33.6  54.9  999.9  1024.3  27.4   9.4   
1   2    Afghanistan  2020-01-23  42.0  32.7  55.9  999.9  1020.8  22.8  14.9   
2   3    Afghanistan  2020-01-24  40.1  36.9  43.2  999.9  1018.6  34.5  10.4   
3   4    Afghanistan  2020-01-25  46.0  37.9  56.3  999.9  1018.0  37.8   6.1   
4   5    Afghanistan  2020-01-26  42.8  36.1  53.1  999.9  1014.8  33.2  10.8   

    prcp  fog  tempaverage monthly temperature (C)  
0   0.00    0                                 35.6  
1  99.99    1                                 35.6  
2   0.17    1                                 35.6  
3   0.57    1                                 35.6  
4   0.00    1                                 35.6

In [54]:
df_weather["Year"] = df_weather["Date"].str.split("-").map(lambda x: x[0])
df_weather["Month"] = df_weather["Date"].str.split("-").map(lambda x: x[1])

In [55]:
df_weather.head()

Id Country_Region        Date  temp   min   max    stp     slp  dewp  wdsp  \
0   1    Afghanistan  2020-01-22  42.6  33.6  54.9  999.9  1024.3  27.4   9.4   
1   2    Afghanistan  2020-01-23  42.0  32.7  55.9  999.9  1020.8  22.8  14.9   
2   3    Afghanistan  2020-01-24  40.1  36.9  43.2  999.9  1018.6  34.5  10.4   
3   4    Afghanistan  2020-01-25  46.0  37.9  56.3  999.9  1018.0  37.8   6.1   
4   5    Afghanistan  2020-01-26  42.8  36.1  53.1  999.9  1014.8  33.2  10.8   

    prcp  fog  tempaverage monthly temperature (C)  Year Month  
0   0.00    0                                 35.6  2020    01  
1  99.99    1                                 35.6  2020    01  
2   0.17    1                                 35.6  2020    01  
3   0.57    1                                 35.6  2020    01  
4   0.00    1                                 35.6  2020    01

In [56]:
df_togo = df_weather[df_weather["Country_Region"] == "Togo"]

In [57]:
df_togo.head()

Id Country_Region        Date  temp   min   max    stp  slp  dewp  \
18174  26563           Togo  2020-01-22  83.2  71.2  96.1  999.9  NaN  52.1   
18175  26564           Togo  2020-01-23  83.4  74.1  96.6  999.9  NaN  46.5   
18176  26565           Togo  2020-01-24  82.9  77.0  95.4  999.9  NaN  33.3   
18177  26566           Togo  2020-01-25  82.3  73.4  95.9  999.9  NaN  30.0   
18178  26567           Togo  2020-01-26  81.4  66.6  95.5  999.9  NaN  30.9   

       wdsp  prcp  fog  tempaverage monthly temperature (C)  Year Month  
18174   1.7   0.0    0                                 84.6  2020    01  
18175   3.3   0.0    0                                 84.6  2020    01  
18176   7.2   0.0    0                                 84.6  2020    01  
18177   6.9   0.0    0                                 84.6  2020    01  
18178   3.5   0.0    0                                 84.6  2020    01

In [58]:
togo_avg_by_month = df_togo[["temp", "Month"]].groupby("Month").mean().sort_values("temp")
togo_avg_by_month

temp
Month           
01     82.220000
04     84.562500
02     84.589655
03     85.390323

In [59]:
togo_avg_by_month.iloc[0,0], togo_avg_by_month.iloc[-1,0]

(82.22, 85.39032258064516)

In [60]:
df_weather[["Country_Region", "temp"]][df_weather["Month"] == "03"].groupby("Country_Region").mean().sort_values("temp").head().index

Index(['Canada', 'Slovakia', 'Russia', 'Mongolia', 'Kazakhstan'], dtype='object', name='Country_Region')

## **<font color='#306998'>TASK </font><font color='#ffd33b'>17</font>**

Select 5 Countries from `training_data_with_weather_info_week_4.csv` and transform the data in such a way that the values `Date` column are not duplicated (i.e. for each date five columns with temperature for the corresponding cities).

In [61]:
countries = df_weather["Country_Region"].unique()[:5]

In [62]:
countries

array(['Afghanistan', 'Albania', 'Algeria', 'Andorra', 'Angola'],
      dtype=object)

In [63]:
df_weather5c = df_weather[["Country_Region", "Date", "temp"]][df_weather["Country_Region"].isin(countries)].groupby(["Country_Region", "Date"]).mean()

In [64]:
df_weather5c = df_weather5c.reset_index()

In [65]:
df_weather5c.head()

Country_Region        Date  temp
0    Afghanistan  2020-01-22  42.6
1    Afghanistan  2020-01-23  42.0
2    Afghanistan  2020-01-24  40.1
3    Afghanistan  2020-01-25  46.0
4    Afghanistan  2020-01-26  42.8

In [66]:
transformed = pd.pivot_table(df_weather5c, index="Date", values="temp", columns="Country_Region")

In [67]:
transformed.head()

Country_Region  Afghanistan  Albania  Algeria  Andorra  Angola
Date                                                          
2020-01-22             42.6     42.3     64.9     26.2    84.3
2020-01-23             42.0     44.1     62.0     27.2    83.6
2020-01-24             40.1     46.7     62.6     28.7    84.0
2020-01-25             46.0     51.5     64.8     27.9    83.1
2020-01-26             42.8     52.5     68.1     28.8    82.7

## **<font color='#306998'>TASK </font><font color='#ffd33b'>18</font>**

Group the data accordingly and determine the average values of minimum, maximum and average daily temperatures for each country and month pair, ie.:

| Country_Region | Month | temp      | min       | max       |
|----------------|-------|-----------|-----------|-----------|
| Afghanistan    | 01    | 33.42     | 27.98     | 40.38     |
| Afghanistan    | 02    | 31.275862 | 22.255172 | 41.437931 |
| Afghanistan    | 03    | 37.8      | 29.4      | 48.945161 |
| Afghanistan    | 04    | 45.275    | 36.7875   | 56.3      |
| Albania        | 01    | 49.4      | 41.83     | 59.04     |
| ...            | ...   | ...       | ...       | ...       |
| Zambia         | 04    | 69.4375   | 59.1      | 81.8375   |
| Zimbabwe       | 01    | 70.95     | 58.86     | 82.25     |
| Zimbabwe       | 02    | 69.989655 | 61.634483 | 78.503448 |
| Zimbabwe       | 03    | 69.170968 | 58.667742 | 80.035484 |
| Zimbabwe       | 04    | 72.3      | 62.6625   | 80.9125   |


In [68]:
df_weather[["Country_Region", "Month", "temp", "min", "max"]].groupby(["Country_Region", "Month"]).mean()

temp        min        max
Country_Region Month                                 
Afghanistan    01     33.420000  27.980000  40.380000
               02     31.275862  22.255172  41.437931
               03     37.800000  29.400000  48.945161
               04     45.275000  36.787500  56.300000
Albania        01     49.400000  41.830000  59.040000
...                         ...        ...        ...
Zambia         04     69.437500  59.100000  81.837500
Zimbabwe       01     70.950000  58.860000  82.250000
               02     69.989655  61.634483  78.503448
               03     69.170968  58.667742  80.035484
               04     72.300000  62.662500  80.912500

[736 rows x 3 columns]